### ISSUE TITLE
LG-50: FAT Approval Test Summary

### ISSUE THREAD

**Issue by fmhwong** [OpenLiberty/open-liberty#7949]
**1) Describe the test strategy & approach for this feature, and describe how the approach verifies the functions delivered by this feature. The description should include the positive and negative testing done, whether all testing is automated, what manual tests exist (if any) and where the tests are stored (source control). Automated testing is expected for all features with manual testing considered an exception to the rule.**

For any feature, be aware that only FAT tests (not unit or BVT) are executed in our cross platform testing. To ensure cross platform testing ensure you have sufficient FAT coverage to verify the feature.

If delivering tests outside of the standard Liberty FAT framework, do the tests push the results into cognitive testing database (if not, consult with the CSI Team who can provide advice and verify if results are being received)?_

**The TCK (FAT) is the comprehensive test coverage for this feature. This was developed in tandem with the Red Hat team to cover the extent of the changes to MicroProfile Metrics from 1.x to 2.x. It should be noted that the TCK extends on top of the MicroProfile Metrics 1.x TCK. It provides a thorough coverage of the use case of the MicroProfile Metrics API usage (Java API + annotation usage).**

**There was manual execution of the TCK throughout development.**

**Additionally there is Metrics 2.0 monitor autofeature that starts up when both `monitor-1.0` feature and `mpMetrics-2.0` feature is enabled. This auto-feature enables an additional set of _vendor_ metrics. A Metrics 2.0 Monitor FAT is introduced to cover these metrics. This FAT is built on top of a metrics 1.x monitor FAT.**

**FAT tests will include the MicroProfile Metrics 2.0 TCK.**
**The TCK FAT bucket is** `com.ibm.ws.microprofile.metrics.2.0_fat_tck`
**The monitor FAT bucket is** `com.ibm.ws.microprofile.metrics.2.0.monitor_fat`

----------------------------------------------------------------------------------------------------------------

**2) Collectively as a team you need to assess your confidence in the testing delivered based on the values below.  This should be done as a team and not an individual to ensure more eyes are on it and that pressures to deliver quickly are absorbed by the team as a whole.**

Please indicate your confidence in the testing (up to and including FAT) delivered with this feature by selecting one of these values:

0 - No automated testing delivered

1 - We have minimal automated coverage of the feature including golden paths.  There is a relatively high risk that defects or issues could be found in this feature.

2 - We have delivered a reasonable automated coverage of the golden paths of this feature but are aware of gaps and extra testing that could be done here.  Error/outlying scenarios are not really covered.  There are likely risks that issues may exist in the golden paths

3 - We have delivered all automated testing we believe is needed for the golden paths of this feature and minimal coverage of the error/outlying scenarios.  There is a risk when the feature is used outside the golden paths however we are confident on the golden path.  Note:  This may still be a valid end state for a feature... things like Beta features may well suffice at this level.

4 - We have delivered all automated testing we believe is needed for the golden paths of this feature and have good coverage of the error/outlying scenarios.  While more testing of the error/outlying scenarios could be added we believe there is minimal risk here and the cost of providing these is considered higher than the benefit they would provide.

5 - We have delivered all automated testing we believe is needed for this feature.  The testing covers all golden path cases as well as all the error/outlying scenarios that make sense.  We are not aware of any gaps in the testing at this time. No manual testing is required to verify this feature.

**4**

Based on your answer above, for any answer other than a 4 or 5 please provide details of what drove your answer.  Please be aware, it may be perfectly reasonable in some scenarios to deliver with any value above.  We may accept no automated testing is needed for some features, we may be happy with low levels of testing on samples for instance so please don't feel the need to drive to a 5.  We need your honest assessment as a team and the reasoning for why you believe shipping at that level is valid. What are the gaps, what is the risk etc.  Please also provide links to the follow on work that is needed to close the gaps (should you deem it needed)

**N/A**

**Comment by dave-waddling** [comment]
**FAT Focal Approval comment:** Hi @fmhwong, thanks for completing this Feature Test Summary and for clarifying a few things via Slack. 

While the TCK has good coverage of the mp-metrics 2.0 feature, is it sufficient to ensure test coverage of Liberty specific interactions (e.g. with other features, logging, etc.)? Please can you update the Feature Test Summary to clarify this.

Finally, I will need to wait for the the results of this weekend's SOE to ensure the testing is passing cleanly before approving; so I'll be hoping to approve on Monday.

**Comment by dave-waddling** [comment]
**FAT Focal Approval comment:** Hi @fmhwong, thanks for the updates to the Feature Test Summary, I also spoke with @Channyboy who also gave me some extra info. Please can you triage the defects below  to identify whether they are test/product issues and add your findings to this issue? 

- [263198] (https://wasrtc.hursley.ibm.com:9443/jazz/resource/itemName/com.ibm.team.workitem.WorkItem/263198) (looks to be intermittent)
- [261878](https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.workitem.viewWorkItem&id=261878) (seen in the SOE and Windows platform builds)
- [262549](https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.workitem.viewWorkItem&id=262549) (looks to be file locking on Windows)
- [260682](https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.workitem.viewWorkItem&id=260682) (intermittent on Windows)

**Comment by Channyboy** [comment]
@dave-waddling 
@fmhwong 

263198 (looks to be intermittent): _test problem_

261878 (seen in the SOE and Windows platform builds): Looks to be a test problem. 

262549 (looks to be file locking on Windows): _test problem_

260682 (intermittent on Windows) - More of a platform issue (slowness)

**Comment by dave-waddling** [comment]
**FAT Focal Approval comment:** Hi @Channyboy, thanks for triaging those defects, it's good to know they're only test bugs. Given that you've triaged them it might be a good idea to add the `test_bug` tag as per the [triage process](https://github.ibm.com/was-liberty/WS-CD-Open/wiki/Liberty-Functional-Areas#functional-area-responsibilities). 

Unfortunately, the first 3 defects in the list, while test bugs, are impacting the builds too much for me to grant a FAT Focal Approval before they are fixed.

**Comment by Channyboy** [comment]
@dave-waddling 
These test issues were _"inherited"_ from the FATs from the mpMetric 1.x fats.
These are not newly introduced.
These defects are currently being looked at by others

**Comment by dave-waddling** [comment]
**FAT Focal Approval comment:** Hi @Channyboy, thanks for pointing that out, you are right that these defects pre-existed this feature but now they are hitting the buckets for both versions of mpMetrics. I think in balance the right thing to do is approve this feature and also make the non-intermittent defects high impact to ensure they are addressed as a priority; I'll add the FAT Focal Approval to the parent Epic.

**Comment by Channyboy** [comment]
For 261878 - (also will be written in the Test Summary)
To summarize, that specific test is missing a vendor metric. This also belongs to the mpMetrics-2.0-monitor-1.0 auto-feature. The root cause is that an mbean  is not able to be retrieved from the mbean server of the JVM. Manual tests shows that starting up a server with the mpMetrics and monitor (which will start up the mpMetrics-2.0-monitor-1.0 autofeature) will get your the `threadpool_default_executor_active_threads`. This leads me to believe its a _timing_ test issue or the an issue with the JVM

**Comment by sabolo** [comment]
Please can you provide an update on the progress of the defects holding up this feature.  Once you believe you have fixed in a release build for them all we can check the following SOE and ensure that the issues have gone away.  Many thanks

**Comment by Channyboy** [comment]
@sabolo 
The fix has gone in last week via:
https://github.com/OpenLiberty/open-liberty/pull/8440

I don't believe the issue occurred this past weekends run.
However, I wasn't too sure if i'm using the right tool on the dashboard to view results.

https://cognitive.hursley.ibm.com/bucketAnalysis.html?bucket=com.ibm.ws.microprofile.metrics.2.0.monitor_fat

**Comment by dave-waddling** [comment]
FAT Focal Approval comment: @Channyboy and I had a brief discussion on Slack about the outstanding defects associated with this feature.

The only product issue that was outstanding (261878) has been resolved and I've confirmed this is the case in the latest SOE.

The remaining defects are either due to file-locking issues on Windows (262549 and 263198) or a tight timing window in the TCK (260682) which is also being seen by non-IBM companies.

Given the file-locking issues are well understood and that the TCK timing issue is currently outside of our control I'm willing to grant approval at this time.
